packages/js-sdk/src/sandbox.ts

```diff
@@ -170,149 +264,253 @@ export class Sandbox extends SandboxBase {
+      "xrandr", { envs: { DISPLAY: this.display } }
+    );
+
+    const match = result.stdout.match(/(\d+x\d+)/);
+    if (match) {
+      const [width, height] = match[1].split('x').map(Number);
+      return { width, height };
+    }
+    return null;
+  }
+
+  private *breakIntoChunks(text: string, n: number): Generator<string> {
+    for (let i = 0; i < text.length; i += n) {
+      yield text.slice(i, i + n);
+    }
+  }
+
+  private quoteString(s: string): string {
+    if (!s) {
+      return "''";
+    }
+
+    if (!/[^\w@%+=:,./-]/.test(s)) {
+      return s;
+    }
+
+    // use single quotes, and put single quotes into double quotes
+    // the string $'b is then quoted as '$'"'"'b'
+    return "'" + s.replace(/'/g, "'\"'\"'") + "'";
   }
 
   /**
    * Write the given text at the current cursor position.
    * @param text - The text to write.
+   * @param chunkSize - The size of each chunk of text to write.
+   * @param delayInMs - The delay between each chunk of text.
    */
-  async write(text: string) {
-    return this.runPyautoguiCode(`pyautogui.write(${JSON.stringify(text)})`)
+  async write(
+    text: string,
+    chunkSize: number = 25,
+    delayInMs: number = 75
+  ): Promise<void> {
+    const chunks = this.breakIntoChunks(text, chunkSize);
+
+    for (const chunk of chunks) {
+      await this.commands.run(
+        `xdotool type --delay ${delayInMs} ${this.quoteString(chunk)}`,
+        { envs: { DISPLAY: this.display } }
+      );
+    }
   }
 
   /**
    * Press a key.
    * @param key - The key to press (e.g. "enter", "space", "backspace", etc.).
    */
-  async press(key: string) {
-    return this.runPyautoguiCode(`pyautogui.press(${JSON.stringify(key)})`)
+  async press(key: string): Promise<void> {
+    await this.commands.run(
+      `xdotool key ${key}`, { envs: { DISPLAY: this.display } }
+    );
   }
 
   /**
    * Press a hotkey.
-   * @param keys - The keys to press (e.g. `hotkey("ctrl", "c")` will press Ctrl+C).
+   * @param key - The key to press (e.g. "ctrl+c").
    */
-  async hotkey(...keys: string[]) {
-    const keysString = keys.map(key => JSON.stringify(key)).join(', ')
-    return this.runPyautoguiCode(`pyautogui.hotkey(${keysString})`)
+  async hotkey(key: string): Promise<void> {
+    await this.press(key);
   }
 
   /**
    * Open a file or a URL in the default application.
-   * Note that you'll need to wait for the application to be opened.
    * @param fileOrUrl - The file or URL to open.
    */
-  async open(fileOrUrl: string) {
-    return this.commands.run(`xdg-open ${fileOrUrl}`, { background: true })
+  async open(fileOrUrl: string): Promise<void> {
+    await this.commands.run(
+      `xdg-open ${fileOrUrl}`, { background: true, envs: { DISPLAY: this.display } }
+    );
+  }
+}
+
+
+class VNCServer {
+  private vncHandle: CommandHandle | null = null;
+  private novncHandle: CommandHandle | null = null;
+  private readonly url: URL;
+  readonly password: string;
```

Comment:
I want a getPassword method that will return either password (when enableStreamAuth is true) or undefined

Reply:
Make private.